sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala

```diff
@@ -95,7 +100,19 @@ class ParquetWriteSupport extends WriteSupport[InternalRow] with Logging {
 
   override def init(configuration: Configuration): WriteContext = {
     val schemaString = configuration.get(ParquetWriteSupport.SPARK_ROW_SCHEMA)
+    val shreddedSchemaString = configuration.get(ParquetWriteSupport.SPARK_VARIANT_SHREDDING_SCHEMA)
     this.schema = StructType.fromString(schemaString)
+    // If shreddingSchemaString is provided, we use that everywhere in the writer, except for
+    // setting the spark schema in the Parquet metadata. If it isn't provided, it means that there
+    // are no shredded Variant columns, so it is identical to this.schema.
+    this.shreddedSchema = if (shreddedSchemaString == null) {
+      this.schema
+    } else {
+      val v = StructType.fromString(shreddedSchemaString)
+      // A bit awkwardly, the schema string doesn't include metadata to identify which struct
```

Comment:
I'm open to suggestions, but from reading the code, it didn't look like the metadata could be passed through the DDL representation, and I couldn't think of an alternative approach to identify the shredding for each Variant field. Technically, for this PR I guess I could have gotten away with constructing the shredded schema in this code directly from the SQLConf, but in the future, I think we'll want to do it based on table properties and/or buffered data in the task, and I don't think it would make sense for that to go here.

Reply:
Shall we use data type JSON string?

Reply:
Oh, thanks. I had looked at the `json` method in StructType, and forgot that the metadata is in StructField, not StructType, so I missed that it is writing the metadata there. Let me clean this up.

Reply:
I updated the PR to clean this up. I also moved the code to add/check the metadata into `SparkShreddingUtils`.